### PR TITLE
retrieval: Only output matching nodes, MatchPath dagspec

### PR DESCRIPTION
**Pull request by magik6k**
Depends on https://github.com/ipld/go-ipld-selector-text-lite/pull/5

todo:
- [x] Use go-ipld-selector-text-lite master when merged
- [x] Test non-path-matching retrieval
- [x] Make sure that ls / cat still works

### CODE REVIEW

**Comment by jennijuju** [comment]
also need https://github.com/ipld/go-ipld-selector-text-lite/pull/5

File: api/types.go [hunk 1]
```diff
@@ -214,6 +214,10 @@ type DagSpec struct {
 	// - when using textselector, the path specifies subtree
 	// - the matched graph must have a single root
 	DataSelector *Selector
+
+	// MatchPath matches the path traversal when DataSelector is a textselector.
+	// Ignored when DataSelector is a json selector and in non-car retrieval
+	MatchPath bool
```
Comment (ribasushi) [inline api/types.go:220]: That's a crummy name... Perhaps `CarExportFullMerkleProof` ?

Comment (magik6k) [reply api/types.go:220]: `ExportMerkleProof`?

File: node/impl/client/client.go [hunk 2]
```diff
@@ -777,13 +778,13 @@ func getDataSelector(dps *api.Selector) (datamodel.Node, error) {
 			ssb := builder.NewSelectorSpecBuilder(basicnode.Prototype.Any)
 
 			selspec, err := textselector.SelectorSpecFromPath(
-				textselector.Expression(*dps),
+				textselector.Expression(*dps), matchPath,
 
 				// URGH - this is a direct copy from https://github.com/filecoin-project/go-fil-markets/blob/v1.12.0/shared/selectors.go#L10-L16
 				// Unable to use it because we need the SelectorSpec, and markets exposes just a reified node
```
Comment (ribasushi) [inline node/impl/client/client.go:784]:
```suggestion
```

Comment no longer applicable

File: api/types.go [hunk 3]
```diff
@@ -214,6 +214,10 @@ type DagSpec struct {
 	// - when using textselector, the path specifies subtree
 	// - the matched graph must have a single root
 	DataSelector *Selector
+
+	// ExportMerkleProof matches the path traversal when DataSelector is a textselector.
+	// Ignored when DataSelector is a json selector and in non-car retrieval
```
Comment (ribasushi) [inline api/types.go:219]:
```suggestion
	// ExportMerkleProof is applicable only when exporting to a CAR file via a path textselector
	// When true, in addition to the selection target, the resulting CAR will contain every block along the
	// path back to, and including the original root
	// When false the resulting CAR contains only the blocks of the target subdag 
```

**Review by ribasushi — APPROVED** [review]
Great stuff!

2 more suggestions, no functional changes

**Pull request merged**